Failing tests in devel after solver removal

I get the following failed tests in the current devel.

I already reported the moma related ones in the deletion PR, but there are some more.

<details>

```
_______________________________________________________________________________________________ TestCobraFluxAnalysis.test_single_gene_deletion_moma_benchmark ________________________________________________________________________________________________
multiprocessing.pool.RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/usr/lib64/python3.6/multiprocessing/pool.py", line 119, in worker
    result = (True, func(*args, **kwds))
  File "/home/cdiener/code/cobrapy/cobra/flux_analysis/deletion.py", line 96, in _gene_deletion_worker
    return _gene_deletion(_model, ids)
  File "/home/cdiener/code/cobrapy/cobra/flux_analysis/deletion.py", line 85, in _gene_deletion
    growth, _ = _reactions_knockouts_with_restore(model, all_reactions)
  File "/home/cdiener/code/cobrapy/cobra/flux_analysis/deletion.py", line 52, in _reactions_knockouts_with_restore
    growth = _get_growth(model)
  File "/home/cdiener/code/cobrapy/cobra/flux_analysis/deletion.py", line 60, in _get_growth
    growth = restore_biomass(model)
  File "/home/cdiener/code/cobrapy/cobra/flux_analysis/deletion.py", line 42, in restore_biomass
    result += model.reactions.get_by_id(k.name).flux * v
AttributeError: 'Integer' object has no attribute 'name'
"""

The above exception was the direct cause of the following exception:

self = <cobra.test.test_flux_analysis.TestCobraFluxAnalysis object at 0x7f5d13043b00>, model = <Model e_coli_core at 0x7f5d13043ba8>, benchmark = <pytest_benchmark.fixture.BenchmarkFixture object at 0x7f5d129229e8>

    def test_single_gene_deletion_moma_benchmark(self, model, benchmark):
        try:
            sutil.get_solver_name(qp=True)
        except sutil.SolverNotFound:
            pytest.skip("no qp support")
    
        genes = ['b0008', 'b0114', 'b2276', 'b1779']
        benchmark(single_gene_deletion, model, gene_list=genes,
>                 method="moma")

cobra/test/test_flux_analysis.py:155: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../.local/lib/python3.6/site-packages/pytest_benchmark/fixture.py:121: in __call__
    return self._raw(function_to_benchmark, *args, **kwargs)
../../.local/lib/python3.6/site-packages/pytest_benchmark/fixture.py:143: in _raw
    duration, iterations, loops_range = self._calibrate_timer(runner)
../../.local/lib/python3.6/site-packages/pytest_benchmark/fixture.py:266: in _calibrate_timer
    duration = runner(loops_range)
../../.local/lib/python3.6/site-packages/pytest_benchmark/fixture.py:86: in runner
    function_to_benchmark(*args, **kwargs)
cobra/flux_analysis/deletion.py:285: in single_gene_deletion
    **kwargs
cobra/flux_analysis/deletion.py:181: in _multi_deletion
    chunksize=chunk_size))
cobra/flux_analysis/deletion.py:167: in extract_knockout_results
    for (growth, ids) in results}
cobra/flux_analysis/deletion.py:165: in <dictcomp>
    return {frozenset(ids): 0.0 if growth and abs(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <multiprocessing.pool.IMapUnorderedIterator object at 0x7f5d13043d68>, timeout = None

    def next(self, timeout=None):
        with self._cond:
            try:
                item = self._items.popleft()
            except IndexError:
                if self._index == self._length:
                    raise StopIteration
                self._cond.wait(timeout)
                try:
                    item = self._items.popleft()
                except IndexError:
                    if self._index == self._length:
                        raise StopIteration
                    raise TimeoutError
    
        success, value = item
        if success:
            return value
>       raise value
E       AttributeError: 'Integer' object has no attribute 'name'

/usr/lib64/python3.6/multiprocessing/pool.py:735: AttributeError
______________________________________________________________________________________________ TestCobraFluxAnalysis.test_single_deletion_linear_moma_benchmark _______________________________________________________________________________________________
multiprocessing.pool.RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/usr/lib64/python3.6/multiprocessing/pool.py", line 119, in worker
    result = (True, func(*args, **kwds))
  File "/home/cdiener/code/cobrapy/cobra/flux_analysis/deletion.py", line 96, in _gene_deletion_worker
    return _gene_deletion(_model, ids)
  File "/home/cdiener/code/cobrapy/cobra/flux_analysis/deletion.py", line 85, in _gene_deletion
    growth, _ = _reactions_knockouts_with_restore(model, all_reactions)
  File "/home/cdiener/code/cobrapy/cobra/flux_analysis/deletion.py", line 52, in _reactions_knockouts_with_restore
    growth = _get_growth(model)
  File "/home/cdiener/code/cobrapy/cobra/flux_analysis/deletion.py", line 60, in _get_growth
    growth = restore_biomass(model)
  File "/home/cdiener/code/cobrapy/cobra/flux_analysis/deletion.py", line 42, in restore_biomass
    result += model.reactions.get_by_id(k.name).flux * v
AttributeError: 'Integer' object has no attribute 'name'
"""

The above exception was the direct cause of the following exception:

self = <cobra.test.test_flux_analysis.TestCobraFluxAnalysis object at 0x7f5d124a27f0>, model = <Model e_coli_core at 0x7f5d124a2908>, benchmark = <pytest_benchmark.fixture.BenchmarkFixture object at 0x7f5d12468588>

    def test_single_deletion_linear_moma_benchmark(self, model, benchmark):
        genes = ['b0008', 'b0114', 'b2276', 'b1779']
        benchmark(single_gene_deletion, model, gene_list=genes,
>                 method="linear moma")

cobra/test/test_flux_analysis.py:160: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../.local/lib/python3.6/site-packages/pytest_benchmark/fixture.py:121: in __call__
    return self._raw(function_to_benchmark, *args, **kwargs)
../../.local/lib/python3.6/site-packages/pytest_benchmark/fixture.py:143: in _raw
    duration, iterations, loops_range = self._calibrate_timer(runner)
../../.local/lib/python3.6/site-packages/pytest_benchmark/fixture.py:266: in _calibrate_timer
    duration = runner(loops_range)
../../.local/lib/python3.6/site-packages/pytest_benchmark/fixture.py:86: in runner
    function_to_benchmark(*args, **kwargs)
cobra/flux_analysis/deletion.py:285: in single_gene_deletion
    **kwargs
cobra/flux_analysis/deletion.py:181: in _multi_deletion
    chunksize=chunk_size))
cobra/flux_analysis/deletion.py:167: in extract_knockout_results
    for (growth, ids) in results}
cobra/flux_analysis/deletion.py:165: in <dictcomp>
    return {frozenset(ids): 0.0 if growth and abs(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <multiprocessing.pool.IMapUnorderedIterator object at 0x7f5d12440128>, timeout = None

    def next(self, timeout=None):
        with self._cond:
            try:
                item = self._items.popleft()
            except IndexError:
                if self._index == self._length:
                    raise StopIteration
                self._cond.wait(timeout)
                try:
                    item = self._items.popleft()
                except IndexError:
                    if self._index == self._length:
                        raise StopIteration
                    raise TimeoutError
    
        success, value = item
        if success:
            return value
>       raise value
E       AttributeError: 'Integer' object has no attribute 'name'

/usr/lib64/python3.6/multiprocessing/pool.py:735: AttributeError
____________________________________________________________________________________________________ TestCobraFluxAnalysis.test_single_gene_deletion_moma _____________________________________________________________________________________________________
multiprocessing.pool.RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/usr/lib64/python3.6/multiprocessing/pool.py", line 119, in worker
    result = (True, func(*args, **kwds))
  File "/home/cdiener/code/cobrapy/cobra/flux_analysis/deletion.py", line 96, in _gene_deletion_worker
    return _gene_deletion(_model, ids)
  File "/home/cdiener/code/cobrapy/cobra/flux_analysis/deletion.py", line 85, in _gene_deletion
    growth, _ = _reactions_knockouts_with_restore(model, all_reactions)
  File "/home/cdiener/code/cobrapy/cobra/flux_analysis/deletion.py", line 52, in _reactions_knockouts_with_restore
    growth = _get_growth(model)
  File "/home/cdiener/code/cobrapy/cobra/flux_analysis/deletion.py", line 60, in _get_growth
    growth = restore_biomass(model)
  File "/home/cdiener/code/cobrapy/cobra/flux_analysis/deletion.py", line 42, in restore_biomass
    result += model.reactions.get_by_id(k.name).flux * v
AttributeError: 'Integer' object has no attribute 'name'
"""

The above exception was the direct cause of the following exception:

self = <cobra.test.test_flux_analysis.TestCobraFluxAnalysis object at 0x7f5d122c1e48>, model = <Model e_coli_core at 0x7f5d122c8a58>

    def test_single_gene_deletion_moma(self, model):
        try:
            sutil.get_solver_name(qp=True)
        except sutil.SolverNotFound:
            pytest.skip("no qp support")
    
        # expected knockouts for textbook model
        growth_dict = {"b0008": 0.87, "b0114": 0.71, "b0116": 0.56,
                       "b2276": 0.11, "b1779": 0.00}
    
        rates = single_gene_deletion(model,
                                     gene_list=growth_dict.keys(),
>                                    method="moma")

cobra/test/test_flux_analysis.py:196: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
cobra/flux_analysis/deletion.py:285: in single_gene_deletion
    **kwargs
cobra/flux_analysis/deletion.py:181: in _multi_deletion
    chunksize=chunk_size))
cobra/flux_analysis/deletion.py:167: in extract_knockout_results
    for (growth, ids) in results}
cobra/flux_analysis/deletion.py:165: in <dictcomp>
    return {frozenset(ids): 0.0 if growth and abs(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <multiprocessing.pool.IMapUnorderedIterator object at 0x7f5d1306ec88>, timeout = None

    def next(self, timeout=None):
        with self._cond:
            try:
                item = self._items.popleft()
            except IndexError:
                if self._index == self._length:
                    raise StopIteration
                self._cond.wait(timeout)
                try:
                    item = self._items.popleft()
                except IndexError:
                    if self._index == self._length:
                        raise StopIteration
                    raise TimeoutError
    
        success, value = item
        if success:
            return value
>       raise value
E       AttributeError: 'Integer' object has no attribute 'name'

/usr/lib64/python3.6/multiprocessing/pool.py:735: AttributeError
_________________________________________________________________________________________________ TestCobraFluxAnalysis.test_single_gene_deletion_linear_moma _________________________________________________________________________________________________
multiprocessing.pool.RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/usr/lib64/python3.6/multiprocessing/pool.py", line 119, in worker
    result = (True, func(*args, **kwds))
  File "/home/cdiener/code/cobrapy/cobra/flux_analysis/deletion.py", line 96, in _gene_deletion_worker
    return _gene_deletion(_model, ids)
  File "/home/cdiener/code/cobrapy/cobra/flux_analysis/deletion.py", line 85, in _gene_deletion
    growth, _ = _reactions_knockouts_with_restore(model, all_reactions)
  File "/home/cdiener/code/cobrapy/cobra/flux_analysis/deletion.py", line 52, in _reactions_knockouts_with_restore
    growth = _get_growth(model)
  File "/home/cdiener/code/cobrapy/cobra/flux_analysis/deletion.py", line 60, in _get_growth
    growth = restore_biomass(model)
  File "/home/cdiener/code/cobrapy/cobra/flux_analysis/deletion.py", line 42, in restore_biomass
    result += model.reactions.get_by_id(k.name).flux * v
AttributeError: 'Integer' object has no attribute 'name'
"""

The above exception was the direct cause of the following exception:

self = <cobra.test.test_flux_analysis.TestCobraFluxAnalysis object at 0x7f5d12fa0278>, model = <Model e_coli_core at 0x7f5d12fa0160>

    def test_single_gene_deletion_linear_moma(self, model):
        # expected knockouts for textbook model
        growth_dict = {"b0008": 0.87, "b0114": 0.76, "b0116": 0.65,
                       "b2276": 0.08, "b1779": 0.00}
    
        df = single_gene_deletion(model, gene_list=growth_dict.keys(),
>                                 method="linear moma")

cobra/test/test_flux_analysis.py:223: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
cobra/flux_analysis/deletion.py:285: in single_gene_deletion
    **kwargs
cobra/flux_analysis/deletion.py:181: in _multi_deletion
    chunksize=chunk_size))
cobra/flux_analysis/deletion.py:167: in extract_knockout_results
    for (growth, ids) in results}
cobra/flux_analysis/deletion.py:165: in <dictcomp>
    return {frozenset(ids): 0.0 if growth and abs(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <multiprocessing.pool.IMapUnorderedIterator object at 0x7f5d126fca58>, timeout = None

    def next(self, timeout=None):
        with self._cond:
            try:
                item = self._items.popleft()
            except IndexError:
                if self._index == self._length:
                    raise StopIteration
                self._cond.wait(timeout)
                try:
                    item = self._items.popleft()
                except IndexError:
                    if self._index == self._length:
                        raise StopIteration
                    raise TimeoutError
    
        success, value = item
        if success:
            return value
>       raise value
E       AttributeError: 'Integer' object has no attribute 'name'

/usr/lib64/python3.6/multiprocessing/pool.py:735: AttributeError
___________________________________________________________________________________________________________ TestSolverBasedModel.test_solver_change ___________________________________________________________________________________________________________

self = <cobra.test.test_solver_model.TestSolverBasedModel object at 0x7f5c945ce710>, model = <Model e_coli_core at 0x7f5c945ce860>

    @pytest.mark.skipif("cplex" not in solvers, reason="need cplex")
    def test_solver_change(self, model):
        solver_id = id(model.solver)
        problem_id = id(model.solver.problem)
>       solution = model.optimize(solution_type=Solution).fluxes
E       TypeError: optimize() got an unexpected keyword argument 'solution_type'

cobra/test/test_solver_model.py:740: TypeError
_______________________________________________________________________________________________ TestSolverBasedModel.test_solver_change_with_optlang_interface ________________________________________________________________________________________________

self = <cobra.test.test_solver_model.TestSolverBasedModel object at 0x7f5c945b4898>, model = <Model e_coli_core at 0x7f5c945b4a58>

    @pytest.mark.skipif("cplex" not in solvers, reason="need cplex")
    def test_solver_change_with_optlang_interface(self, model):
        solver_id = id(model.solver)
        problem_id = id(model.solver.problem)
>       solution = model.optimize(solution_type=Solution).fluxes
E       TypeError: optimize() got an unexpected keyword argument 'solution_type'

cobra/test/test_solver_model.py:753: TypeError
=============================================================================================== 6 failed, 347 passed, 51 skipped, 14 xfailed in 102.89 seconds ================================================================================================
```

</details>

--- a/cobra/flux_analysis/deletion.py
+++ b/cobra/flux_analysis/deletion.py
@@ -1,6 +1,5 @@
 # -*- coding: utf-8 -*-
 
-import sys
 import multiprocessing
 import logging
 import optlang
@@ -10,7 +9,6 @@ from functools import partial
 from builtins import (map, dict)
 
 import pandas as pd
-from future.utils import raise_
 
 from cobra.manipulation.delete import find_gene_knockout_reactions
 import cobra.util.solver as sutil
@@ -24,7 +22,7 @@ def _reactions_knockouts_with_restore(model, reactions):
         for reaction in reactions:
             reaction.knock_out()
         growth = _get_growth(model)
-    return ([r.id for r in reactions], growth, model.solver.status)
+    return [r.id for r in reactions], growth, model.solver.status
 
 
 def _get_growth(model):
@@ -89,7 +87,7 @@ def _multi_deletion(model, entity, element_lists, method="fba", num_jobs=None):
         List of iterables ``cobra.Reaction``s or ``cobra.Gene``s (or their IDs)
         to be deleted.
 
-    method: {"fba", "moma", "linear-moma"}, optional
+    method: {"fba", "moma", "linear moma"}, optional
         Method used to predict the growth rate.
 
     num_jobs : int, optional
@@ -197,7 +195,7 @@ def single_reaction_deletion(model, reaction_list=None, method="fba",
         ``cobra.Reaction``s to be deleted. If not passed,
         all the reactions from the model are used.
 
-    method: {"fba", "moma", "linear-moma"}, optional
+    method: {"fba", "moma", "linear moma"}, optional
         Method used to predict the growth rate.
 
     num_jobs : int, optional
@@ -238,7 +236,7 @@ def single_gene_deletion(model, gene_list=None, method="fba", num_jobs=None):
         ``cobra.Gene``s to be deleted. If not passed,
         all the genes from the model are used.
 
-    method: {"fba", "moma", "linear-moma"}, optional
+    method: {"fba", "moma", "linear moma"}, optional
         Method used to predict the growth rate.
 
     num_jobs : int, optional
@@ -268,7 +266,9 @@ def single_gene_deletion(model, gene_list=None, method="fba", num_jobs=None):
 def double_reaction_deletion(model, reaction_list1=None, reaction_list2=None,
                              method="fba", num_jobs=None):
     """
-    Knock out each reaction pair from two given lists.
+    Knock out each reaction pair from the combinations of two given lists.
+
+    We say 'pair' here but the order order does not matter.
 
     Parameters
     ----------
@@ -281,10 +281,9 @@ def double_reaction_deletion(model, reaction_list1=None, reaction_list2=None,
 
     reaction_list2 : iterable, optional
         Second iterable of ``cobra.Reaction``s to be deleted. If not passed,
-        all the reactions from the model are used. The order independent
-        pairs of reactions will be used to perform the knockouts.
+        all the reactions from the model are used.
 
-    method: {"fba", "moma", "linear-moma"}, optional
+    method: {"fba", "moma", "linear moma"}, optional
         Method used to predict the growth rate.
 
     num_jobs : int, optional
@@ -318,7 +317,9 @@ def double_reaction_deletion(model, reaction_list1=None, reaction_list2=None,
 def double_gene_deletion(model, gene_list1=None, gene_list2=None,
                          method="fba", num_jobs=None):
     """
-    Knock out each gene pair from two given lists.
+    Knock out each gene pair from the combination of two given lists.
+
+    We say 'pair' here but the order order does not matter.
 
     Parameters
     ----------
@@ -331,10 +332,9 @@ def double_gene_deletion(model, gene_list1=None, gene_list2=None,
 
     gene_list2 : iterable, optional
         Second iterable of ``cobra.Gene``s to be deleted. If not passed,
-        all the genes from the model are used. The order independent
-        pairs of genes will be used to perform the knockouts.
+        all the genes from the model are used.
 
-    method: {"fba", "moma", "linear-moma"}, optional
+    method: {"fba", "moma", "linear moma"}, optional
         Method used to predict the growth rate.
 
     num_jobs : int, optional

--- a/cobra/flux_analysis/deletion.py
+++ b/cobra/flux_analysis/deletion.py
@@ -75,8 +75,7 @@ def _init_worker(model):
 
 def _multi_deletion(model, entity, element_lists, method="fba", num_jobs=None):
     """
-    Helper function that provides the common interface for sequential
-    knockouts
+    Provide a common interface for single or multiple knockouts.
 
     Parameters
     ----------
@@ -84,31 +83,32 @@ def _multi_deletion(model, entity, element_lists, method="fba", num_jobs=None):
         The metabolic model to perform deletions in.
 
     entity : 'gene' or 'reaction'
-        The entity to knockout (`cobra.Gene` or `cobra.Reaction`)
+        The entity to knockout (``cobra.Gene`` or ``cobra.Reaction``).
 
     element_lists : list
-        List of iterables `cobra.Reaction`s or `cobra.Gene`s (or their IDs)
+        List of iterables ``cobra.Reaction``s or ``cobra.Gene``s (or their IDs)
         to be deleted.
 
     method: {"fba", "moma", "linear-moma"}, optional
-        Procedure used to predict the growth rate.
+        Method used to predict the growth rate.
 
     num_jobs : int, optional
-        Number of parallel processes to run. Can speed up the computations
-        if number of knockouts to perform is large. If not passed,
-        will be set to the number of CPUs
+        The number of parallel processes to run. Can speed up the computations
+        if the number of knockouts to perform is large. If not passed,
+        will be set to the number of CPUs found.
 
     Returns
     -------
     pandas.DataFrame
-        A sparse representation of all combinations of entities
-        deletions. The columns are ['ids', 'growth', 'status'], where
-        ids : frozenset([str])
-            all the knockouts performed for one case
+        A representation of all combinations of entity deletions. The
+        columns are 'growth' and 'status', where
+
+        index : frozenset([str])
+            The gene or reaction identifiers that were knocked out.
         growth : float
-            growth rate of optimized model
+            The growth rate of the adjusted model.
         status : str
-            solver status
+            The solution's status.
     """
     solver = sutil.interface_to_str(model.problem.__name__)
     if "moma" in method and solver not in sutil.qp_solvers:
@@ -154,6 +154,7 @@ def _multi_deletion(model, entity, element_lists, method="fba", num_jobs=None):
                 chunksize=chunk_size
             ))
             pool.close()
+            pool.join()
         else:
             worker = dict(gene=_gene_deletion,
                           reaction=_reaction_deletion)[entity]
@@ -184,7 +185,8 @@ def _element_lists(entities, *ids):
 
 def single_reaction_deletion(model, reaction_list=None, method="fba",
                              num_jobs=None):
-    """Sequentially knocks out each reaction from a given reaction list.
+    """
+    Knock out each reaction from a given list.
 
     Parameters
     ----------
@@ -192,28 +194,30 @@ def single_reaction_deletion(model, reaction_list=None, method="fba",
         The metabolic model to perform deletions in.
 
     reaction_list : iterable
-        Iterable `cobra.Reaction`s to be deleted. If not passed,
+        ``cobra.Reaction``s to be deleted. If not passed,
         all the reactions from the model are used.
 
     method: {"fba", "moma", "linear-moma"}, optional
-        Procedure used to predict the growth rate.
+        Method used to predict the growth rate.
 
     num_jobs : int, optional
-        Number of parallel processes to run. Can speed up the computations
-        if number of knockouts to perform is large. If not passed,
-        will be set to the number of CPUs
+        The number of parallel processes to run. Can speed up the computations
+        if the number of knockouts to perform is large. If not passed,
+        will be set to the number of CPUs found.
 
     Returns
     -------
     pandas.DataFrame
-        A sparse representation of all combinations of entities
-        deletions. The columns are ['ids', 'growth', 'status'], where
-        ids : frozenset([str])
-            all the knockouts performed for one case
+        A representation of all single reaction deletions. The columns are
+        'growth' and 'status', where
+
+        index : frozenset([str])
+            The reaction identifier that was knocked out.
         growth : float
-            growth rate of optimized model
+            The growth rate of the adjusted model.
         status : str
-            solver status
+            The solution's status.
+
     """
     return _multi_deletion(
         model, 'reaction',
@@ -222,7 +226,8 @@ def single_reaction_deletion(model, reaction_list=None, method="fba",
 
 
 def single_gene_deletion(model, gene_list=None, method="fba", num_jobs=None):
-    """Sequentially knocks out each gene from a given gene list.
+    """
+    Knock out each gene from a given list.
 
     Parameters
     ----------
@@ -230,28 +235,30 @@ def single_gene_deletion(model, gene_list=None, method="fba", num_jobs=None):
         The metabolic model to perform deletions in.
 
     gene_list : iterable
-        Iterable `cobra.Gene`s to be deleted. If not passed,
+        ``cobra.Gene``s to be deleted. If not passed,
         all the genes from the model are used.
 
     method: {"fba", "moma", "linear-moma"}, optional
-        Procedure used to predict the growth rate.
+        Method used to predict the growth rate.
 
     num_jobs : int, optional
-        Number of parallel processes to run. Can speed up the computations
-        if number of knockouts to perform is large. If not passed,
-        will be set to the number of CPUs
+        The number of parallel processes to run. Can speed up the computations
+        if the number of knockouts to perform is large. If not passed,
+        will be set to the number of CPUs found.
 
     Returns
     -------
     pandas.DataFrame
-        A sparse representation of all combinations of entities
-        deletions. The columns are ['ids', 'growth', 'status'], where
-        ids : frozenset([str])
-            all the knockouts performed for one case
+        A representation of all single gene deletions. The columns are
+        'growth' and 'status', where
+
+        index : frozenset([str])
+            The gene identifier that was knocked out.
         growth : float
-            growth rate of optimized model
+            The growth rate of the adjusted model.
         status : str
-            solver status
+            The solution's status.
+
     """
     return _multi_deletion(
         model, 'gene', element_lists=_element_lists(model.genes, gene_list),
@@ -261,7 +268,7 @@ def single_gene_deletion(model, gene_list=None, method="fba", num_jobs=None):
 def double_reaction_deletion(model, reaction_list1=None, reaction_list2=None,
                              method="fba", num_jobs=None):
     """
-    Sequentially delete pairs of reactions and record the objective value.
+    Knock out each reaction pair from two given lists.
 
     Parameters
     ----------
@@ -269,33 +276,35 @@ def double_reaction_deletion(model, reaction_list1=None, reaction_list2=None,
         The metabolic model to perform deletions in.
 
     reaction_list1 : iterable, optional
-        First iterable of `cobra.Reaction`s to be deleted. If not passed,
+        First iterable of ``cobra.Reaction``s to be deleted. If not passed,
         all the reactions from the model are used.
 
     reaction_list2 : iterable, optional
-        Second iterable of `cobra.Reaction`s to be deleted. If not passed,
-        all the reactions from the model are used. The product of two
-        reactions lists will be used to perform the knockouts.
+        Second iterable of ``cobra.Reaction``s to be deleted. If not passed,
+        all the reactions from the model are used. The order independent
+        pairs of reactions will be used to perform the knockouts.
 
     method: {"fba", "moma", "linear-moma"}, optional
-        Procedure used to predict the growth rate.
+        Method used to predict the growth rate.
 
     num_jobs : int, optional
-        Number of parallel processes to run. Can speed up the computations
-        if number of knockouts to perform is large. If not passed,
-        will be set to the number of CPUs
+        The number of parallel processes to run. Can speed up the computations
+        if the number of knockouts to perform is large. If not passed,
+        will be set to the number of CPUs found.
 
     Returns
     -------
     pandas.DataFrame
-        A sparse representation of all combinations of entities
-        deletions. The columns are ['ids', 'growth', 'status'], where
-        ids : frozenset([str])
-            all the knockouts performed for one case
+        A representation of all combinations of reaction deletions. The
+        columns are 'growth' and 'status', where
+
+        index : frozenset([str])
+            The reaction identifiers that were knocked out.
         growth : float
-            growth rate of optimized model
+            The growth rate of the adjusted model.
         status : str
-            solver status
+            The solution's status.
+
     """
 
     reaction_list1, reaction_list2 = _element_lists(model.reactions,
@@ -308,7 +317,8 @@ def double_reaction_deletion(model, reaction_list1=None, reaction_list2=None,
 
 def double_gene_deletion(model, gene_list1=None, gene_list2=None,
                          method="fba", num_jobs=None):
-    """Sequentially knocks out pairs of genes in a model
+    """
+    Knock out each gene pair from two given lists.
 
     Parameters
     ----------
@@ -316,33 +326,35 @@ def double_gene_deletion(model, gene_list1=None, gene_list2=None,
         The metabolic model to perform deletions in.
 
     gene_list1 : iterable, optional
-        First iterable of `cobra.Gene`s to be deleted. If not passed,
+        First iterable of ``cobra.Gene``s to be deleted. If not passed,
         all the genes from the model are used.
 
     gene_list2 : iterable, optional
-        Second iterable of `cobra.Gene`s to be deleted. If not passed,
-        all the genes from the model are used. The product of two
-        genes lists will be used to perform the knockouts.
+        Second iterable of ``cobra.Gene``s to be deleted. If not passed,
+        all the genes from the model are used. The order independent
+        pairs of genes will be used to perform the knockouts.
 
     method: {"fba", "moma", "linear-moma"}, optional
-        Procedure used to predict the growth rate.
+        Method used to predict the growth rate.
 
     num_jobs : int, optional
-        Number of parallel processes to run. Can speed up the computations
-        if number of knockouts to perform is large. If not passed,
-        will be set to the number of CPUs
+        The number of parallel processes to run. Can speed up the computations
+        if the number of knockouts to perform is large. If not passed,
+        will be set to the number of CPUs found.
 
     Returns
     -------
     pandas.DataFrame
-        A sparse representation of all combinations of entities
-        deletions. The columns are ['ids', 'growth', 'status'], where
-        ids : frozenset([str])
-            all the knockouts performed for one case
+        A representation of all combinations of gene deletions. The
+        columns are 'growth' and 'status', where
+
+        index : frozenset([str])
+            The gene identifiers that were knocked out.
         growth : float
-            growth rate of optimized model
+            The growth rate of the adjusted model.
         status : str
-            solver status
+            The solution's status.
+
     """
 
     gene_list1, gene_list2 = _element_lists(model.genes, gene_list1,

--- a/cobra/flux_analysis/deletion.py
+++ b/cobra/flux_analysis/deletion.py
@@ -4,45 +4,19 @@ import sys
 import multiprocessing
 import logging
 import optlang
-import math
-from six import iteritems
 from warnings import warn
 from itertools import product
-from collections import defaultdict
 from functools import partial
 from builtins import (map, dict)
+
+import pandas as pd
 from future.utils import raise_
+
 from cobra.manipulation.delete import find_gene_knockout_reactions
 import cobra.util.solver as sutil
-from cobra.flux_analysis import moma
+from cobra.flux_analysis.moma import add_moma
 
 LOGGER = logging.getLogger(__name__)
-
-
-def infinite_defaultdict():
-    return defaultdict(infinite_defaultdict)
-
-
-def defaultdict_to_dict(def_dict):
-    if not isinstance(def_dict, defaultdict):
-        return def_dict
-    for k, v in iteritems(def_dict):
-        def_dict[k] = defaultdict_to_dict(v)
-    return dict(def_dict)
-
-
-def biomass_reaction(model):
-    return model.objective.expression.as_coefficients_dict()
-
-
-def restore_biomass(model):
-    result = 0
-    for k, v in iteritems(model.notes['biomass_reaction']):
-        try:
-            result += model.reactions.get_by_id(k.name).flux * v
-        except KeyError:
-            pass
-    return result
 
 
 def _reactions_knockouts_with_restore(model, reactions):
@@ -50,20 +24,18 @@ def _reactions_knockouts_with_restore(model, reactions):
         for reaction in reactions:
             reaction.knock_out()
         growth = _get_growth(model)
-    return (growth, [r.id for r in reactions])
+    return ([r.id for r in reactions], growth, model.solver.status)
 
 
 def _get_growth(model):
     try:
         if 'moma_old_objective' in model.solver.variables:
             model.slim_optimize()
-            growth = restore_biomass(model)
+            growth = model.solver.variables.moma_old_objective.primal
         else:
             growth = model.slim_optimize()
-        if math.isnan(growth):
-            growth = None
     except optlang.exceptions.SolverError:
-        growth = None
+        growth = float('nan')
     return growth
 
 
@@ -82,8 +54,8 @@ def _gene_deletion(model, ids):
                 model, (model.genes.get_by_id(g_id),)
             )
         )
-    growth, _ = _reactions_knockouts_with_restore(model, all_reactions)
-    return (growth, ids)
+    _, growth, status = _reactions_knockouts_with_restore(model, all_reactions)
+    return (ids, growth, status)
 
 
 def _reaction_deletion_worker(ids):
@@ -101,101 +73,93 @@ def _init_worker(model):
     _model = model
 
 
-def _multi_deletion(cobra_model, entity, element_lists, method="fba",
-                    number_of_processes=None, solver=None,
-                    zero_cutoff=1e-12):
+def _multi_deletion(model, entity, element_lists, method="fba", num_jobs=None):
     """
     Helper function that provides the common interface for sequential
     knockouts
 
     Parameters
     ----------
-    cobra_model : cobra.Model
+    model : cobra.Model
         The metabolic model to perform deletions in.
 
+    entity : 'gene' or 'reaction'
+        The entity to knockout (`cobra.Gene` or `cobra.Reaction`)
+
     element_lists : list
-        List of lists `cobra.Reaction`s or `cobra.Gene`s (or their IDs)
+        List of iterables `cobra.Reaction`s or `cobra.Gene`s (or their IDs)
         to be deleted.
 
-    method: {"fba", "moma"}, optional
+    method: {"fba", "moma", "linear-moma"}, optional
         Procedure used to predict the growth rate.
 
-    solver: str, optional
-        This must be a QP-capable solver for MOMA. If left unspecified,
-        a suitable solver will be automatically chosen.
-
-    zero_cutoff: float, optional
-        When checking to see if a value is 0, this threshold is used.
+    num_jobs : int, optional
+        Number of parallel processes to run. Can speed up the computations
+        if number of knockouts to perform is large. If not passed,
+        will be set to the number of CPUs
 
     Returns
     -------
-    dict of dict
-        A sparse representation of all combinations of pairs of reaction
-        deletions without replacements.
+    pandas.DataFrame
+        A sparse representation of all combinations of entities
+        deletions. The columns are ['ids', 'growth', 'status'], where
+        ids : frozenset([str])
+            all the knockouts performed for one case
+        growth : float
+            growth rate of optimized model
+        status : str
+            solver status
     """
-    with cobra_model as model:
+    solver = sutil.interface_to_str(model.problem.__name__)
+    if "moma" in method and solver not in sutil.qp_solvers:
+        raise RuntimeError(
+            "Cannot use MOMA since '{}' is not QP-capable."
+            "Please choose a different solver or use FBA only.".format(solver))
+
+    if num_jobs is None:
         try:
-            solver = sutil.choose_solver(model, solver, qp=(method == "moma"))
-        except sutil.SolverNotFound:
-            if method == "moma":
-                warn(
-                    "Cannot use MOMA since no QP-capable solver was found. "
-                    "Falling back to FBA.")
-                solver = sutil.choose_solver(model, solver)
-            else:
-                (err_type, err_val, err_tb) = sys.exc_info()
-                raise_(err_type, err_val, err_tb)  # reraise for Python2&3
+            num_cpu = multiprocessing.cpu_count()
+        except NotImplementedError:
+            warn("Number of cores could not be detected - assuming 1.")
+            num_cpu = 1
+    else:
+        num_cpu = num_jobs
 
-        if number_of_processes is None:
-            try:
-                num_cpu = multiprocessing.cpu_count()
-            except NotImplementedError:
-                warn("Number of cores could not be detected - assuming 1.")
-                num_cpu = 1
-        else:
-            num_cpu = number_of_processes
-
-        if 'moma' in method:
-            model.notes['biomass_reaction'] = biomass_reaction(model)
-            moma.add_moma(model, linear='linear' in method)
+    with model:
+        if "moma" in method:
+            add_moma(model, linear="linear" in method)
 
         args = set([frozenset(comb) for comb in product(*element_lists)])
+        num_cpu = min(num_cpu, len(args))
 
-        def extract_knockout_results(results):
-            return {frozenset(ids): 0.0 if growth and abs(
-                growth) < zero_cutoff else growth
-                    for (growth, ids) in results}
+        def extract_knockout_results(result_iter):
+            result = pd.DataFrame([
+                (frozenset(ids), growth, status)
+                for (ids, growth, status) in result_iter
+            ], columns=['ids', 'growth', 'status'])
+            result.set_index('ids', inplace=True)
+            return result
 
         if num_cpu > 1:
             num_cpu = min(num_cpu, len(args))
-            WORKER_FUNCTIONS = dict(
-                gene=_gene_deletion_worker,
-                reaction=_reaction_deletion_worker
-            )
+            worker = dict(gene=_gene_deletion_worker,
+                          reaction=_reaction_deletion_worker)[entity]
             chunk_size = len(args) // num_cpu
             pool = multiprocessing.Pool(
                 num_cpu, initializer=_init_worker, initargs=(model,)
             )
-            results = extract_knockout_results(
-                pool.imap_unordered(WORKER_FUNCTIONS[entity], args,
-                                    chunksize=chunk_size))
+            results = extract_knockout_results(pool.imap_unordered(
+                worker,
+                args,
+                chunksize=chunk_size
+            ))
             pool.close()
         else:
-            WORKER_FUNCTIONS = dict(
-                gene=_gene_deletion,
-                reaction=_reaction_deletion
-            )
+            worker = dict(gene=_gene_deletion,
+                          reaction=_reaction_deletion)[entity]
             results = extract_knockout_results(map(
-                partial(WORKER_FUNCTIONS[entity], model), args
-            ))
-        double = infinite_defaultdict()
-        for comb in product(*element_lists):
-            growth = results[frozenset(comb)]
-            current = double
-            for v in comb[:-1]:
-                current = current[v]
-            current[comb[-1]] = growth
-        return defaultdict_to_dict(double)
+                partial(worker, model), args))
+        return results
 
 
 def _entities_ids(entities):
@@ -218,153 +182,171 @@ def _element_lists(entities, *ids):
     return result
 
 
-def single_reaction_deletion(model, reaction_list=None, **kwargs):
+def single_reaction_deletion(model, reaction_list=None, method="fba",
+                             num_jobs=None):
     """Sequentially knocks out each reaction from a given reaction list.
 
     Parameters
     ----------
-    model : a cobra model
-        The model from which to delete the reactions. The model will not be
-        modified.
+    model : cobra.Model
+        The metabolic model to perform deletions in.
+
     reaction_list : iterable
-        List of reaction IDs or cobra.Reaction. If None (default) will use all
-        reactions in the model.
-    method : str, optional
-        The method used to obtain fluxes. Must be one of "fba" or "moma".
-    solver : str, optional
-        Name of the solver to be used.
-    solver_args : optional
-        Additional arguments for the solver. Ignored for optlang solver, please
-        use `model.solver.configuration` instead.
+        Iterable `cobra.Reaction`s to be deleted. If not passed,
+        all the reactions from the model are used.
+
+    method: {"fba", "moma", "linear-moma"}, optional
+        Procedure used to predict the growth rate.
+
+    num_jobs : int, optional
+        Number of parallel processes to run. Can speed up the computations
+        if number of knockouts to perform is large. If not passed,
+        will be set to the number of CPUs
 
     Returns
     -------
-    tuple of 2 dictionaries
-        The first dictionary maps each reaction id to its growth rate after
-        the knockout. The second tuple reports the solutions status (for
-        instance "optimal" for each knockout).
+    pandas.DataFrame
+        A sparse representation of all combinations of entities
+        deletions. The columns are ['ids', 'growth', 'status'], where
+        ids : frozenset([str])
+            all the knockouts performed for one case
+        growth : float
+            growth rate of optimized model
+        status : str
+            solver status
     """
     return _multi_deletion(
-        model,
-        'reaction',
+        model, 'reaction',
         element_lists=_element_lists(model.reactions, reaction_list),
-        **kwargs
-    )
+        method=method, num_jobs=num_jobs)
 
 
-def single_gene_deletion(model, gene_list=None, **kwargs):
+def single_gene_deletion(model, gene_list=None, method="fba", num_jobs=None):
     """Sequentially knocks out each gene from a given gene list.
 
     Parameters
     ----------
-    cobra_model : a cobra model
-        The model from which to delete the genes. The model will not be
-        modified.
+    model : cobra.Model
+        The metabolic model to perform deletions in.
+
     gene_list : iterable
-        List of gene IDs or cobra.Gene. If None (default) will use all genes in
-        the model.
-    method : str, optional
-        The method used to obtain fluxes. Must be one of "fba" or "moma".
-    solver : str, optional
-        Name of the solver to be used.
-    solver_args : optional
-        Additional arguments for the solver. Ignored for optlang solver, please
-        use `model.solver.configuration` instead.
+        Iterable `cobra.Gene`s to be deleted. If not passed,
+        all the genes from the model are used.
+
+    method: {"fba", "moma", "linear-moma"}, optional
+        Procedure used to predict the growth rate.
+
+    num_jobs : int, optional
+        Number of parallel processes to run. Can speed up the computations
+        if number of knockouts to perform is large. If not passed,
+        will be set to the number of CPUs
 
     Returns
     -------
-    tuple of 2 dictionaries
-        The first dictionary maps each gene id to its growth rate after
-        the knockout. The second tuple reports the solutions status (for
-        instance "optimal" for each knockout).
+    pandas.DataFrame
+        A sparse representation of all combinations of entities
+        deletions. The columns are ['ids', 'growth', 'status'], where
+        ids : frozenset([str])
+            all the knockouts performed for one case
+        growth : float
+            growth rate of optimized model
+        status : str
+            solver status
     """
     return _multi_deletion(
-        model,
-        'gene',
-        element_lists=_element_lists(model.genes, gene_list),
-        **kwargs
-    )
+        model, 'gene', element_lists=_element_lists(model.genes, gene_list),
+        method=method, num_jobs=num_jobs)
 
 
-def double_reaction_deletion(model,
-                             reaction_list1=None, reaction_list2=None,
-                             **kwargs):
+def double_reaction_deletion(model, reaction_list1=None, reaction_list2=None,
+                             method="fba", num_jobs=None):
     """
     Sequentially delete pairs of reactions and record the objective value.
 
     Parameters
     ----------
     model : cobra.Model
-        The metabolic model to perform double deletions in.
+        The metabolic model to perform deletions in.
 
-    reaction_list1, reaction_list2 : list, optional
-        Lists of `cobra.Reaction`s (or their IDs) to be deleted. If not
-        provided, all reactions in the model sorted by ID will be used.
-        `reaction_list2` will use `reaction_list1` by default.
+    reaction_list1 : iterable, optional
+        First iterable of `cobra.Reaction`s to be deleted. If not passed,
+        all the reactions from the model are used.
 
-    method: {"fba", "moma"}, optional
+    reaction_list2 : iterable, optional
+        Second iterable of `cobra.Reaction`s to be deleted. If not passed,
+        all the reactions from the model are used. The product of two
+        reactions lists will be used to perform the knockouts.
+
+    method: {"fba", "moma", "linear-moma"}, optional
         Procedure used to predict the growth rate.
 
-    solver: str, optional
-        This must be a QP-capable solver for MOMA. If left unspecified,
-        a suitable solver will be automatically chosen.
-
-    zero_cutoff: float, optional
-        When checking to see if a value is 0, this threshold is used.
+    num_jobs : int, optional
+        Number of parallel processes to run. Can speed up the computations
+        if number of knockouts to perform is large. If not passed,
+        will be set to the number of CPUs
 
     Returns
     -------
-    dict of dict
-        A sparse representation of all combinations of pairs of reaction
-        deletions without replacements.
+    pandas.DataFrame
+        A sparse representation of all combinations of entities
+        deletions. The columns are ['ids', 'growth', 'status'], where
+        ids : frozenset([str])
+            all the knockouts performed for one case
+        growth : float
+            growth rate of optimized model
+        status : str
+            solver status
     """
 
     reaction_list1, reaction_list2 = _element_lists(model.reactions,
                                                     reaction_list1,
                                                     reaction_list2)
-    return _multi_deletion(model, 'reaction',
-                           element_lists=[reaction_list1, reaction_list2],
-                           **kwargs)
+    return _multi_deletion(
+        model, 'reaction', element_lists=[reaction_list1, reaction_list2],
+        method=method, num_jobs=num_jobs)
 
 
-def double_gene_deletion(model, gene_list1=None, gene_list2=None, **kwargs):
+def double_gene_deletion(model, gene_list1=None, gene_list2=None,
+                         method="fba", num_jobs=None):
     """Sequentially knocks out pairs of genes in a model
 
-    model : :class:`~cobra.core.Model.Model`
-        cobra model in which to perform deletions
+    Parameters
+    ----------
+    model : cobra.Model
+        The metabolic model to perform deletions in.
 
-    gene_list1 : [:class:`~cobra.core.Gene.Gene`:] (or their id's)
-        Genes to be deleted. These will be the keys in the resulting dict.
-        If not provided, all genes will be used.
+    gene_list1 : iterable, optional
+        First iterable of `cobra.Gene`s to be deleted. If not passed,
+        all the genes from the model are used.
 
-    gene_list1 : [:class:`~cobra.core.Gene.Gene`:] (or their id's)
-        Genes to be deleted. These will be the keys in the resulting dict.
-        If not provided, gene_list1 will be used.
+    gene_list2 : iterable, optional
+        Second iterable of `cobra.Gene`s to be deleted. If not passed,
+        all the genes from the model are used. The product of two
+        genes lists will be used to perform the knockouts.
 
-    method: "fba" or "moma"
-        Procedure used to predict the growth rate
+    method: {"fba", "moma", "linear-moma"}, optional
+        Procedure used to predict the growth rate.
 
-    solver: str for solver name
-        This must be a QP-capable solver for MOMA. If left unspecified,
-        a suitable solver will be automatically chosen.
+    num_jobs : int, optional
+        Number of parallel processes to run. Can speed up the computations
+        if number of knockouts to perform is large. If not passed,
+        will be set to the number of CPUs
 
-    zero_cutoff: float
-        When checking to see if a value is 0, this threshold is used.
-
-    number_of_processes: int for number of processes to use.
-        If unspecified, the number of parallel processes to use will be
-        automatically determined. Setting this to 1 explicitly disables used
-        of the multiprocessing library.
-
-    .. note:: multiprocessing is not supported with method=moma
-
-    return_frame: bool
-        If true, formats the results as a pandas.Dataframe. Otherwise
-        returns a dict of the form:
-        {"x": row_labels, "y": column_labels", "data": 2D matrix}
+    Returns
+    -------
+    pandas.DataFrame
+        A sparse representation of all combinations of entities
+        deletions. The columns are ['ids', 'growth', 'status'], where
+        ids : frozenset([str])
+            all the knockouts performed for one case
+        growth : float
+            growth rate of optimized model
+        status : str
+            solver status
     """
 
     gene_list1, gene_list2 = _element_lists(model.genes, gene_list1,
                                             gene_list2)
-    return _multi_deletion(model, 'gene',
-                           element_lists=[gene_list1, gene_list2], **kwargs)
+    return _multi_deletion(
+        model, 'gene', element_lists=[gene_list1, gene_list2],
+        method=method, num_jobs=num_jobs)

--- a/cobra/flux_analysis/variability.py
+++ b/cobra/flux_analysis/variability.py
@@ -2,10 +2,13 @@
 
 from __future__ import absolute_import
 
-import pandas
-from optlang.symbolics import Zero
+import math
 from warnings import warn
 from itertools import chain
+
+import pandas
+from optlang.symbolics import Zero
+from six import iteritems
 
 from cobra.flux_analysis.loopless import loopless_fva_iter
 from cobra.flux_analysis.parsimonious import add_pfba
@@ -212,7 +215,7 @@ def find_blocked_reactions(model, reaction_list=None,
                 max(abs(min_max)) < zero_cutoff]
 
 
-def find_essential_genes(model):
+def find_essential_genes(model, threshold=None):
     """Return a set of essential genes.
 
     A gene is considered essential if restricting the flux of all reactions
@@ -223,26 +226,24 @@ def find_essential_genes(model):
     ----------
     model : cobra.Model
         The model to find the essential genes for.
+    threshold : float, optional
+        Minimal objective flux to be considered viable. By default this is
+        0.01 times the growth rate.
 
     Returns
     -------
     set
         Set of essential genes
     """
-    solution = model.optimize(raise_error=True)
-    tolerance = model.solver.configuration.tolerances.feasibility
-    non_zero_flux_reactions = list(
-        solution[solution.fluxes.abs() > tolerance].index)
-    genes_to_check = set(chain.from_iterable(
-        model.reactions.get_by_id(rid).genes for rid in
-        non_zero_flux_reactions))
-    deletions = single_gene_deletion(model, gene_list=genes_to_check,
-                                     method='fba')
-    gene_ids = [k for k, v in deletions.items() if not v]
-    return set(model.genes.get_by_any(gene_ids))
+    if threshold is None:
+        threshold = model.slim_optimize(error_value=None) * 1E-02
+    deletions = single_gene_deletion(model, method='fba')
+    essential = deletions.loc[deletions['growth'].isna() |
+                              (deletions['growth'] < threshold), :].index
+    return set(model.genes.get_by_id(g) for ids in essential for g in ids)
 
 
-def find_essential_reactions(model):
+def find_essential_reactions(model, threshold=None):
     """Return a set of essential reactions.
 
     A reaction is considered essential if restricting its flux to zero
@@ -252,18 +253,18 @@ def find_essential_reactions(model):
     ----------
     model : cobra.Model
         The model to find the essential reactions for.
+    threshold : float, optional
+        Minimal objective flux to be considered viable. By default this is
+        0.01 times the growth rate.
 
     Returns
     -------
     set
         Set of essential reactions
     """
-    solution = model.optimize(raise_error=True)
-    tolerance = model.solver.configuration.tolerances.feasibility
-    non_zero_flux_reactions = list(
-        solution[solution.fluxes.abs() > tolerance].index)
-    deletions = single_reaction_deletion(model,
-                                         reaction_list=non_zero_flux_reactions,
-                                         method='fba')
-    reaction_ids = [k for k, v in deletions.items() if not v]
-    return set(model.reactions.get_by_any(reaction_ids))
+    if threshold is None:
+        threshold = model.slim_optimize(error_value=None) * 1E-02
+    deletions = single_reaction_deletion(model, method='fba')
+    essential = deletions.loc[deletions['growth'].isna() |
+                              (deletions['growth'] < threshold), :].index
+    return set(model.reactions.get_by_id(r) for ids in essential for r in ids)

--- a/cobra/test/test_flux_analysis.py
+++ b/cobra/test/test_flux_analysis.py
@@ -5,10 +5,10 @@ from __future__ import absolute_import
 import re
 import sys
 import warnings
-from contextlib import contextmanager
-
+import math
 import pytest
 import numpy
+from contextlib import contextmanager
 from optlang.interface import OPTIMAL, INFEASIBLE
 from six import StringIO, iteritems
 
@@ -28,6 +28,7 @@ optlang_solvers = ["optlang-" + s for s in stable_optlang if s in
                    sutil.solvers]
 all_solvers = ["optlang-" + s for s in stable_optlang if
                s in sutil.solvers]
+qp_solvers = ["cplex"]
 
 
 def construct_ll_test_model():
@@ -140,13 +141,14 @@ class TestCobraFluxAnalysis:
                        "b2276": 0.21, "b1779": 0.00}
         rates = single_gene_deletion(model,
                                      gene_list=growth_dict.keys(),
-                                     method="fba")
+                                     method="fba")["growth"]
         for gene, expected_value in iteritems(growth_dict):
-            assert abs(rates[gene] - expected_value) < 0.01
+            assert abs(rates[frozenset([gene])] - expected_value) < 0.01
 
     def test_single_gene_deletion_moma_benchmark(self, model, benchmark):
         try:
-            sutil.get_solver_name(qp=True)
+            solver = sutil.get_solver_name(qp=True)
+            model.solver = solver
         except sutil.SolverNotFound:
             pytest.skip("no qp support")
 
@@ -155,6 +157,12 @@ class TestCobraFluxAnalysis:
                   method="moma")
 
     def test_single_deletion_linear_moma_benchmark(self, model, benchmark):
+        try:
+            solver = sutil.get_solver_name(qp=True)
+            model.solver = solver
+        except sutil.SolverNotFound:
+            pytest.skip("no qp support")
+
         genes = ['b0008', 'b0114', 'b2276', 'b1779']
         benchmark(single_gene_deletion, model, gene_list=genes,
                   method="linear moma")
@@ -162,7 +170,8 @@ class TestCobraFluxAnalysis:
     def test_moma_sanity(self, model):
         """Test optimization criterion and optimality."""
         try:
-            sutil.get_solver_name(qp=True)
+            solver = sutil.get_solver_name(qp=True)
+            model.solver = solver
         except sutil.SolverNotFound:
             pytest.skip("no qp support")
 
@@ -183,7 +192,8 @@ class TestCobraFluxAnalysis:
 
     def test_single_gene_deletion_moma(self, model):
         try:
-            sutil.get_solver_name(qp=True)
+            solver = sutil.get_solver_name(qp=True)
+            model.solver = solver
         except sutil.SolverNotFound:
             pytest.skip("no qp support")
 
@@ -191,14 +201,19 @@ class TestCobraFluxAnalysis:
         growth_dict = {"b0008": 0.87, "b0114": 0.71, "b0116": 0.56,
                        "b2276": 0.11, "b1779": 0.00}
 
-        rates = single_gene_deletion(model,
-                                     gene_list=growth_dict.keys(),
-                                     method="moma")
+        result = single_gene_deletion(
+            model, gene_list=growth_dict.keys(), method="moma")["growth"]
         for gene, expected_value in iteritems(growth_dict):
-            assert abs(rates[gene] - expected_value) < 0.01
+            assert abs(result[frozenset([gene])] - expected_value) < 0.01
 
     def test_linear_moma_sanity(self, model):
         """Test optimization criterion and optimality."""
+        try:
+            solver = sutil.get_solver_name(qp=True)
+            model.solver = solver
+        except sutil.SolverNotFound:
+            pytest.skip("no qp support")
+
         sol = model.optimize()
         with model:
             model.reactions.PFK.knock_out()
@@ -215,13 +230,20 @@ class TestCobraFluxAnalysis:
         assert moma_sabs < sabs
 
     def test_single_gene_deletion_linear_moma(self, model):
+        try:
+            solver = sutil.get_solver_name(qp=True)
+            model.solver = solver
+        except sutil.SolverNotFound:
+            pytest.skip("no qp support")
+
         # expected knockouts for textbook model
         growth_dict = {"b0008": 0.87, "b0114": 0.76, "b0116": 0.65,
                        "b2276": 0.08, "b1779": 0.00}
 
-        df = single_gene_deletion(model, gene_list=growth_dict.keys(),
-                                  method="linear moma")
-        assert all(abs(df[gene] - expected) < 0.01
+        result = single_gene_deletion(
+            model, gene_list=growth_dict.keys(),
+            method="linear moma")['growth']
+        assert all(abs(result[frozenset([gene])] - expected) < 0.01
                    for gene, expected in iteritems(growth_dict))
         with model:
             add_moma(model, linear=True)
@@ -241,9 +263,9 @@ class TestCobraFluxAnalysis:
 
         model.solver = solver
         results = single_reaction_deletion(
-            model, reaction_list=expected_results.keys())
-        for reaction, value in results.items():
-            assert abs(value - expected_results[reaction]) < 0.00001
+            model, reaction_list=expected_results.keys()).to_dict()['growth']
+        for reaction, value in iteritems(expected_results):
+            assert abs(results[frozenset([reaction])] - value) < 1E-05
 
     @classmethod
     def compare_matrices(cls, matrix1, matrix2, places=3):
@@ -314,30 +336,27 @@ class TestCobraFluxAnalysis:
                                  'b1276': 0.863,
                                  'b2935': 0.863,
                                  'b4025': 0.863}}
-        solution = double_gene_deletion(model, gene_list1=genes,
-                                        number_of_processes=4)
-        solution_one_process = double_gene_deletion(model, gene_list1=genes,
-                                                    number_of_processes=1)
-        for (rxn_a, sub) in growth_dict.items():
-            for rxn_b, growth in sub.items():
-                assert round(solution[rxn_a][rxn_b], 3) == growth, \
-                    "unexpected {}-{} double knock-out growth".format(rxn_a,
-                                                                      rxn_b)
-                assert round(solution_one_process[rxn_a][rxn_b],
-                             3) == growth, \
-                    "unexpected {}-{} double knock-out growth".format(rxn_a,
-                                                                      rxn_b)
+        solution = double_gene_deletion(
+            model, gene_list1=genes, num_jobs=4)['growth']
+        solution_one_process = double_gene_deletion(
+            model, gene_list1=genes, num_jobs=1)['growth']
+        for (rxn_a, sub) in iteritems(growth_dict):
+            for rxn_b, growth in iteritems(sub):
+                sol = solution[frozenset([rxn_a, rxn_b])]
+                sol_one = solution_one_process[frozenset([rxn_a, rxn_b])]
+                assert round(sol, 3) == growth
+                assert round(sol_one, 3) == growth
 
     def test_double_reaction_deletion(self, model):
         reactions = ['FBA', 'ATPS4r', 'ENO', 'FRUpts2']
         growth_dict = {
             "FBA": {
                 "ATPS4r": 0.135,
-                "ENO": None,
+                "ENO": float('nan'),
                 "FRUpts2": 0.704
             },
             "ATPS4r": {
-                "ENO": None,
+                "ENO": float('nan'),
                 "FRUpts2": 0.374
             },
             "ENO": {
@@ -345,31 +364,20 @@ class TestCobraFluxAnalysis:
             },
         }
 
-        solution = double_reaction_deletion(model,
-                                            reaction_list1=reactions,
-                                            number_of_processes=4)
+        solution = double_reaction_deletion(
+            model, reaction_list1=reactions, num_jobs=3)['growth']
         solution_one_process = double_reaction_deletion(
-                model,
-                reaction_list1=reactions,
-                number_of_processes=1
-            )
-        for (rxn_a, sub) in growth_dict.items():
-            for rxn_b, growth in sub.items():
-                if growth is None:
-                    assert solution[rxn_a][rxn_b] is growth, \
-                        "unexpected {}-{} double knock-out growth".format(
-                            rxn_a, rxn_b)
-                    assert solution[rxn_a][rxn_b] is growth, \
-                        "unexpected {}-{} double knock-out growth".format(
-                            rxn_a, rxn_b)
+            model, reaction_list1=reactions, num_jobs=1)['growth']
+        for (rxn_a, sub) in iteritems(growth_dict):
+            for rxn_b, growth in iteritems(sub):
+                sol = solution[frozenset([rxn_a, rxn_b])]
+                sol_one = solution_one_process[frozenset([rxn_a, rxn_b])]
+                if math.isnan(growth):
+                    assert math.isnan(sol)
+                    assert math.isnan(sol_one)
                 else:
-                    assert round(solution[rxn_a][rxn_b], 3) == growth, \
-                        "unexpected {}-{} double knock-out growth".format(
-                            rxn_a, rxn_b)
-                    assert round(solution_one_process[rxn_a][rxn_b],
-                                 3) == growth, \
-                        "unexpected {}-{} double knock-out growth".format(
-                            rxn_a, rxn_b)
+                    assert round(sol, 3) == growth
+                    assert round(sol_one, 3) == growth
 
     def test_double_reaction_deletion_benchmark(self, large_model,
                                                 benchmark):

--- a/cobra/test/test_solver_model.py
+++ b/cobra/test/test_solver_model.py
@@ -737,27 +737,12 @@ class TestSolverBasedModel:
     def test_solver_change(self, model):
         solver_id = id(model.solver)
         problem_id = id(model.solver.problem)
-        solution = model.optimize(solution_type=Solution).fluxes
+        solution = model.optimize().fluxes
         model.solver = "cplex"
         assert id(model.solver) != solver_id
         assert id(model.solver.problem) != problem_id
-        new_solution = model.optimize(solution_type=Solution).fluxes
-        for key in list(solution.keys()):
-            assert round(abs(new_solution[key] - solution[key]),
-                         7) == 0
-
-    @pytest.mark.skipif("cplex" not in solvers, reason="need cplex")
-    def test_solver_change_with_optlang_interface(self, model):
-        solver_id = id(model.solver)
-        problem_id = id(model.solver.problem)
-        solution = model.optimize(solution_type=Solution).fluxes
-        model.solver = optlang.cplex_interface
-        assert id(model.solver) != solver_id
-        assert id(model.solver.problem) != problem_id
-        new_solution = model.optimize(solution_type=Solution).fluxes
-        for key in list(solution.keys()):
-            assert round(abs(new_solution[key] - solution[key]),
-                         7) == 0
+        new_solution = model.optimize().fluxes
+        assert numpy.allclose(solution, new_solution, rtol=0, atol=1E-06)
 
     def test_no_change_for_same_solver(self, model):
         solver_id = id(model.solver)

